### PR TITLE
Security considerations

### LINKED ISSUE #11: zero current implementations of Turtle/TriG/N-Triples/N-Quads functionality are spec compliant, because none can "ensure that malignant strings may not be used to mislead the reader" — there's just no way to do so!
_Originally posted by @TallTed in https://github.com/w3c/rdf-concepts/pull/7#discussion_r1092328636_

```
    Application rendering strings retrieved from untrusted RDF documents must ensure
    that malignant strings may not be used to mislead the reader.
```
>>> @[TallTed](https://github.com/TallTed) [18 hours ago](https://github.com/w3c/rdf-concepts/pull/7#discussion_r1091284122)
>>>
>>> I don't believe this requirement can be satisfied by any applications available today, if ever.
>>>
>>> As things stand, humans must take care to load varyingly trustworthy RDF into appropriately partitioned data stores (typically, using named graphs), and to execute queries across "all data in the store" only when recognizing that results may include malicious, erroneous, or otherwise undesirable statements (whether triples or quads).
>>>
>>> Such undesirable results are not the responsibility of any software in the stack.
>>
>> @[gkellogg](https://github.com/gkellogg) [18 hours ago](https://github.com/w3c/rdf-concepts/pull/7#discussion_r1091294960)
>>
>> It's taken right out of the [Turtle IANA Considerations section](https://www.w3.org/TR/turtle/#sec-mediaReg)
>>
>> Turtle can express data which is presented to the user, for example, RDF Schema labels. Application rendering strings retrieved from untrusted Turtle documents must ensure that malignant strings may not be used to mislead the reader. The security considerations in the media type registration for XML ([[RFC3023](https://www.w3.org/TR/turtle/#bib-RFC3023)] section 10) provide additional guidance around the expression of arbitrary data and markup.
>>
>> I didn't want to try to tread new ground here, but ReSpec demands that specs have such a section, and this seemed to apply generally to RDF serializations. I also don't want to revisit the IANA registrations of Turtle/TriG/N-Triples/N-Quads which already have such considerations, and it would be inconsistent to say something different.
>
> @[TallTed](https://github.com/TallTed) [now](https://github.com/w3c/rdf-concepts/pull/7#discussion_r1092328636)
>
> Wow. As I understand things, that means that zero current implementations of Turtle/TriG/N-Triples/N-Quads functionality are spec compliant, because none can "ensure that malignant strings may not be used to mislead the reader" — there's just no way to do so!

### PR DESCRIPTION
**Pull request by gkellogg**
* Update security considerations based on work in N-Quads. Fixes #11.
* Update issue on IRI to #15.
* Add informative section on the rdf:JSON datatype for #14.

***
<a href="https://pr-preview.s3.amazonaws.com/w3c/rdf-concepts/pull/16.html" title="Last updated on Mar 14, 2023, 8:59 PM UTC (40a2a74)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/rdf-concepts/16/1f16ce9...40a2a74.html" title="Last updated on Mar 14, 2023, 8:59 PM UTC (40a2a74)">Diff</a>

### CODE REVIEW

File: spec/index.html
Comment (TallTed) [inline spec/index.html:1146]: ```The `rdf:JSON` is``` should be either ```The `rdf:JSON` datatype was``` or ``` `rdf:JSON` was```

File: spec/index.html
Comment (TallTed) [inline spec/index.html:1324]: The wording in the other instance of this warning is better. "prevent ... from being used to mislead" remains impossible, and SHOULD is no better than MUST for an impossibility.

**Review by TallTed — CHANGES_REQUESTED** [review]
Most are minor. `SHOULD prevent` is little better than `MUST prevent`, as that prevention remains an impossibility, and we should get away from specifying such an impossibility sooner than later.

File: spec/index.html
Comment (gkellogg) [inline spec/index.html:1146]:
```suggestion
      The `rdf:JSON`datatype  is originally defined in
```

File: spec/index.html
Comment (gkellogg) [reply spec/index.html:1324]: How about the following:

```suggestion
    SHOULD ensure that malignant strings may not be used to mislead the reader.
```

`SHOULD` seems better than `must`, even though the later is rendered to be non-normative. But, it seems that as this is a normative section, statements such as this should use normative wording.

**Review by pfps — CHANGES_REQUESTED** [review]
This PR makes substantive changes that need discussion in the WG.

**Comment by domel** [comment]
One does not exclude the other. I don't see why blocking this PR, there is a corresponding box with information in places for WG discussions.

**Comment by pfps** [comment]
I don't think the WG has even voted to take up this sort of change to RDF.

**Comment by gkellogg** [comment]
> This PR makes substantive changes that need discussion in the WG.

Can you be more specific? It includes issue markers, and extracts the security considerations from the mime type to its own section, which is called for in publishing standards. We have been working to correct issues in the security consideration section to address various issues. This text is used in a number of specs. Do you think the full WG needs to resolve to make changes to security considerations?

issue markers are not normative changes, but serve as cues to readers about issues under discussion that may lead to changes in future versions.

File: spec/index.html
Comment (afs) [inline spec/index.html:495]: This isn't a fair balance of the situation. 

Simply state that the WG is aware RFC3987 has not progressed (avoid "abandoned").

In general, I don't think these notes should present only one choice nor should assume the first suggestion is the direction of the working group.

File: spec/index.html
Comment (afs) [inline spec/index.html:500]: The FPWD is from the working group. Do not suggest this is the direction of the working group.

**Review by afs — CHANGES_REQUESTED** [review]

**Comment by pfps** [comment]
There are three separate changes in this PR.   Security, IRIs, and JSON.  I'm not aware that the WG has decided to consider any of them.   I don't see any open issues at https://github.com/w3c/rdf-star-wg/issues where these changes could fit.

**Comment by gkellogg** [comment]
> There are three separate changes in this PR. Security, IRIs, and JSON. I'm not aware that the WG has decided to consider any of them. I don't see any open issues at https://github.com/w3c/rdf-star-wg/issues where these changes could fit.

Generally, these are related to errata, or the consequences of dealing with the errata.

* Security Considerations is an issue for spec-track documents, as is Privacy Considerations. This is more a matter of moving text which involves making reasonable updates.
* IRIs have been problematic due to the lack of progress on RFC3987 when prevents it from being normatively cited. We need to do something here, and issue #15 is open to try to figure that out. In the mean time, an issue marker tells the community that it's being considered.
* `rdf:JSON` is not an erratum, but it's been considered good practice to put issue markers in drafts about open issues.

It sounds like you think that the chairs need to set a policy on the scope of editorial changes related to dealing with errata and updated requirements for spec-track documents.

**Comment by TallTed** [comment]
> It sounds like you think that the chairs need to set a policy on the scope of editorial changes related to dealing with errata and updated requirements for spec-track documents.

Without speaking to the specifics of this PR #16 or the discussion on it, I *would* suggest that keeping each PR (and each preceding Issue) to a single topic — by which I mean a single error when addressing errata from the previous spec — will be a better practice than mushing several topics/errors together, as is being done here.

File: spec/index.html
Comment (TallTed) [inline spec/index.html:1141]:
```suggestion
      The `rdf:JSON` datatype  is originally defined in
```

**Comment by gkellogg** [comment]
Generally, adding issue markers for open issues has been considered good practice, at least in groups I've been involved with. When doing a PR keeping the active issues listed in the spec was considered useful. However, here it seems to have become contentious.

I've removed the updated wording on the #15 and new issue marker for #14 to focus solely on the update to security considerations.

**Comment by TallTed** [comment]
> Generally, adding issue markers for open issues has been considered good practice, at least in groups I've been involved with. When doing a PR keeping the active issues listed in the spec was considered useful. However, here it seems to have become contentious.

If this PR were just adding issue markers for open issues, I think there would have been no contention.  However, it was adding those markers *and* making other changes to the content, which mix was unsurprisingly (to me) contentious...

**Comment by afs** [comment]
The issue markers take the description from the issue.

But initial descriptions have been containing opinion and proposal and then do not reflect dicussion in the issue.

Here we have "Update issue on IRI to https://github.com/w3c/rdf-concepts/issues/15." Nothing to do with security considerations.

**Comment by TallTed** [comment]
@afs wrote:
> The issue markers take the description from the issue.
>
> But initial descriptions have been containing opinion and proposal and then do not reflect dicussion in the issue.
>
> Here we have "Update issue on IRI to https://github.com/w3c/rdf-concepts/issues/15." Nothing to do with security considerations.

Yes.  This is another reason why (as I've noted elsewhere) I would prefer there to be one concern per Issue, and one Issue per PR, and to have meaningful naming and description of both Issue and PR.  The initial Issue description need not be revised to reflect subsequent discussion, so long as it is a simple description of the concern with neither personal opinion nor proposed solution.

File: spec/index.html
Comment (TallTed) [inline spec/index.html:1310]: Such "ensurance" remains impossible, whether `SHOULD` or `MUST`, and this guidance should be changed to the same wording I've suggested elsewhere. If such rewording is not acceptable by the Editors' at this point, I request that an Issue be raised for this specific warning, so that we can work toward WG consensus on what the appropriate guidance is, and apply that to all documents now containing the impossible `SHOULD` or `MUST`.

**Review by TallTed — CHANGES_REQUESTED** [review]

File: spec/index.html
Comment (afs) [reply spec/index.html:496]: We could take this out.

As [pointed out](https://github.com/w3c/rdf-concepts/issues/15#issuecomment-1436613178), many RFC are "Proposed Standard" (including URNs).

Comment (TallTed) [reply spec/index.html:496]: Yes. I noted this removal option in another review of the same basic text in another copy of this shared-yet-not-shared document/content. (I have generally been focused on editing the text in place, rather than the larger meaning of that text.)

Perhaps we need a distinct repo for the shared/common content, from which all that shared content is imported by whatever means. I know that one experiment failed because it broke previews. I think there's another experiment in progress. I hope that or another like it bears fruit soon!

File: spec/index.html
Comment (afs) [inline spec/index.html:1310]:
```suggestion
    SHOULD ensure that strings may not be used to mislead the reader.
```
(I thought I'd suggested this before but that seems t have got lost)

Comment (TallTed) [reply spec/index.html:1310]: How is any app supposed to review all strings and ensure that they are not misleading, any more than malignant?

As I've said in at least one thread on at least one repo (but I think it's several at this point), warnings and other mechanisms _may_ be appropriate, but it's unclear to me whether there's a good place for any of these to be displayed.

Comment (afs) [reply spec/index.html:1310]: The suggestion is to remove the word. Nothing more. it is unnecessary and adds nothing because the text after explains the intent to mislead.

Comment (TallTed) [reply spec/index.html:1310]: Yes, I understand your suggestion. My point is that the advisory remains impossible without "malignant", so I don't see the point of removing that word without changing the rest of the sentence.

File: spec/index.html
Comment (gkellogg) [reply spec/index.html]: Removed.

File: spec/index.html
Comment (gkellogg) [reply spec/index.html:1305]: We have parallel text here at in https://github.com/w3c/rdf-n-quads/pull/17. @TallTed suggested changes there, which I've applied. If we can focus on the N-Quads text until fully approved, we can synchronized those changes back here, and ultimately to the other affected specifications.

Comment (TallTed) [reply spec/index.html:1305]: The relevant text is (currently) at insert lines 329-332 of PR #17.

**Comment by gkellogg** [comment]
@afs and @pfps, you both have outstanding changes requested that prevent merging this PR. Can you please either approve, or suggest additional changes?

This version should closely match that in https://github.com/w3c/rdf-n-quads/pull/19 which has approvals.

**Review by afs — APPROVED** [review]

**Review by TallTed — APPROVED** [review]

**Review by pfps — APPROVED** [review]
The changes here now appear to be much better limited.  The only problem is that the commit list has a lot of other stuff in it.   It would be better to squash all the commits into one when merging or squash locally and force-push.

**Comment by gkellogg** [comment]
My practice has been to do squash merges when there are more than a couple of commits.

A force push may become necessary if there are any conflicts with other PRs, but probably the time to quash is when ultimately merging the PR.

**Pull request merged**
